src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2857,6 +2857,579 @@ ep_rt_mono_write_event_ee_startup_start (void)
 		NULL);
 }
 
+#define STACK_ALLOC 256
+
+// The maximum number of type parameters for a BulkTypeValue instance
+// Aligned with coreCLR StackSArray<ULONGLONG> rgTypeParameters
+static const uint32_t k_size_of_type_parameter_array = STACK_ALLOC / sizeof(intptr_t);
+
+// !!!!!!! NOTE !!!!!!!!
+// The flags must match those in the ETW manifest exactly
+// !!!!!!! NOTE !!!!!!!!
+
+typedef enum {
+	K_ETW_TYPE_FLAGS_DELEGATE = 0x1,
+	K_ETW_TYPE_FLAGS_FINALIZABLE = 0x2,
+	K_ETW_TYPE_FLAGS_EXTERNALLY_IMPLEMENTED_COM_OBJECT = 0x4,
+	K_ETW_TYPE_FLAGS_ARRAY = 0x8,
+
+	K_ETW_TYPE_FLAGS_ARRAY_RANK_MASK = 0x3F00,
+	K_ETW_TYPE_FLAGS_ARRAY_RANK_SHIFT = 8,
+	K_ETW_TYPE_FLAGS_ARRAY_RANK_MAX = K_ETW_TYPE_FLAGS_ARRAY_RANK_MASK >> K_ETW_TYPE_FLAGS_ARRAY_RANK_SHIFT
+} EtwTypeFlags;
+
+// This only contains the fixed-size data at the top of each struct in
+// the bulk type event.  These fields must still match exactly the initial
+// fields of the struct described in the manifest.
+typedef struct _EventStructBulkTypeFixedSizedData {
+	uint64_t type_id;
+	uint64_t module_id;
+	uint32_t type_name_id;
+	uint32_t flags;
+	uint8_t cor_element_type;
+} EventStructBulkTypeFixedSizedData;
+
+// Represents one instance of the Value struct inside a single BulkType event
+typedef struct _BulkTypeValue {
+	EventStructBulkTypeFixedSizedData fixed_sized_data;
+	uint32_t c_type_parameters;
+	intptr_t rg_type_parameters[k_size_of_type_parameter_array];
+	MonoType *rg_mono_type_parameters[k_size_of_type_parameter_array];
+	ep_char8_t *s_name;
+} BulkTypeValue;
+
+// Clear out BulkTypeValue before filling it out (array elements can get reused if there
+// are enough types that we need to flush to multiple events).
+static
+void
+ep_rt_bulk_type_value_init(BulkTypeValue *bulk_type_value)
+{
+	bulk_type_value->fixed_sized_data.type_id = 0;
+	bulk_type_value->fixed_sized_data.module_id = 0;
+	bulk_type_value->fixed_sized_data.type_name_id = 0;
+	bulk_type_value->fixed_sized_data.flags = 0;
+	bulk_type_value->fixed_sized_data.cor_element_type = 0;
+	bulk_type_value->c_type_parameters = 0;
+	memset(bulk_type_value->rg_type_parameters, 0, k_size_of_type_parameter_array);
+	memset(bulk_type_value->rg_mono_type_parameters, 0, k_size_of_type_parameter_array);
+	if (bulk_type_value->s_name)
+		bulk_type_value->s_name[0] = '\0';
```

Comment:
I don't understand what this is supposed to do.

The first time that `ep_rt_bulk_type_value_init` is called, the `s_name` field is NULL, so nothing happens.

When you later call it to re-initialize the bulk type value after it's been fired, this apparently overwrites the old string value's data?

That doesn't seem like what you want.  If you are storing a pointer to the name that is in some other data structure, you just corrupted someone else's data.

If you're storing a string that you allocated yourself, then you should free it (see my earlier suggestion about a destroy function) and set `s_name` to null.   You don't know how much storage you allocated for `s_name`, so if you're hoping to reuse the buffer you don't have enough info.

Keep it simple: initialize `s_name` to null. always store a copy of the name.  destroy the bulk type value after it's been flushed and deallocate `s_name` and reset it to NULL.

Reply:
Since we are keeping `s_name` null for now, removed all instances of setting the first byte to `\0`